MathiasMagnus SAXPY tutorial

neon60 opened the pull request:
The SAXPY tutorial.

randyh62 commented on docs/tutorials/reduction.rst:
in C++ it's been signidfied by...

Spelling needs to be corrected, though I would change it to "in C++ it is called..."

neon60 replied:
Fixed

randyh62 commented on docs/tutorials/reduction.rst:
doesn not

neon60 replied:
Fixed

randyh62 commented on docs/tutorials/reduction.rst:
One such is what ``std::reduce`` does, is to require the user-provided binary...

The approach taken by ``std::reduce`` is to require the user-provided binary...

neon60 replied:
Fixed

randyh62 commented on docs/tutorials/reduction.rst:
Implementing reductions on GPUs will require a basic...

requires a basic...

neon60 replied:
Fixed

randyh62 commented on docs/tutorials/reduction.rst:
through the {ref}`inherent_thread_model`, and

This is not the correct reference syntax for RST. It does not render correctly in the output.

neon60 replied:
Fixed.

randyh62 commented on docs/tutorials/reduction.rst:
Synchronizing _all_ the threads running on a GPU at any given time, while
possible, is a costly and/or intricate operation.

neon60 replied:
Fixed.

randyh62 commented on docs/tutorials/reduction.rst:
We'll be exploring

You'll be exploring...generally we should avoid the use of we.

neon60 replied:
Move the reduction to a separate PR.

randyh62 commented on docs/tutorials/reduction.rst:
If not absolutely necessary, we'll map our parallel algorithm as such that Multi Processors / Blocks can make independent progress and need not sync often.

If synchronization is not absolutely necessary, map the parallel algorithm so that multi-processors and blocks can make independent progress and need not sync often.

I don't think Blocks needs to be capitalized in the text.

neon60 replied:
Fixed.

randyh62 commented on docs/tutorials/reduction.rst:
One may reduce divergence by keeping dataflow

You can reduce divergence...

neon60 replied:
Changed.

randyh62 commented on docs/tutorials/reduction.rst:
the coming code segments

the following code segments...

neon60 replied:
Fixed.

randyh62 commented on docs/tutorials/reduction.rst:
strictly* lockstep

Not sure what the * is needed for

neon60 replied:
Fixed.

randyh62 commented on docs/tutorials/reduction.rst:
Like we've mentioned in the previous step,

As mentioned in the prior step,

neon60 replied:
Fixed.

randyh62 reviewed (commented):
provided comments

neon60 commented:
``IMAGE OF FINAL ALGO`` is missing

neon60 commented:
> `IMAGE OF FINAL ALGO` is missing

Fixed.

peterjunpark commented:
Added language suggestions from @lpaoletti. Review the diff if needed [2b4a8be](https://github.com/ROCm/HIP/pull/3487/commits/2b4a8beb8de2fea01faebd7cf911a61fed9cc578)

neon60 commented:
@lpaoletti Can I get an approval?

neon60 commented:
The reduction tutorial PR: #3513

peterjunpark commented:
Should we remove the tab sync?

I don't like how the page jumps around when you select a tab due to the length of content changing all over the place at the same time.

neon60 commented:
> Should we remove the tab sync?
> 
> I don't like how the page jumps around when you select a tab due to content lengths changing everywhere at the same time.

Probably we should. I am thinking what we can do with it.

randyh62 commented:
> > Should we remove the tab sync?
> > I don't like how the page jumps around when you select a tab due to content lengths changing everywhere at the same time.
> 
> Probably we should. I am thinking what we can do with it.

I think we should keep it. I think a user is going to pick their flow in the first tab, and the page will update the content to display their flow. I don't think people will be flipping back and forth between flows except to see what differences there might be.

lpaoletti commented on docs/tutorial/saxpy.rst:
```diff
@@ -0,0 +1,786 @@
+Heterogeneous programming
+=========================
+
+*Heterogeneous programming* and *offloading APIs* are often said in the same
```
Heterogeneous programming and offloading APIs are often mentioned together.

peterjunpark commented on docs/tutorial/saxpy.rst:
```diff
@@ -0,0 +1,786 @@
+=========================
+
+*Heterogeneous programming* and *offloading APIs* are often said in the same
+breath. Heterogeneous programming deals with devices of varying capabilities
```
```suggestion
*Heterogeneous programming* and *offloading APIs* are often mentioned together. Heterogeneous programming deals with devices of varying capabilities
```

cjatin commented on docs/tutorial/saxpy.rst:
```diff
@@ -0,0 +1,780 @@
+*******************************************************************************
+
+This tutorial explains the basic concepts of the single-source
+Heterogeneous-computing Interface for Portability (HIP) programming model and
```
Can we change it to HIP (Heterogeneous-computing Interface) programming model. I think most people will call it HIP instead of its full name.

neon60 replied:
This is defined by our documentation styling guide, and we are using the same at other locations. "The full text and then the acronym in parenthesis, and the following references use the acronym."

Example:
https://github.com/ROCm/HIP/blob/develop/docs/index.md

cjatin commented on docs/tutorial/saxpy.rst:
```diff
@@ -0,0 +1,780 @@
+============================
+
+First, let's do the "Hello, World!" of GPGPU: SAXPY. *SAXPY* is a mathematical
+acronym; a vector equation :math:`a\cdot x+y=z` where :math:`a\in\mathbb{R}` is
```
I think you have mentioned SAXPY's full form below. Can we move it up here.

neon60 replied:
Added.

cjatin commented on docs/tutorial/saxpy.rst:
```diff
@@ -0,0 +1,780 @@
+
+``HIP_CHECK`` is a custom macro borrowed from the examples utilities which
+checks the error code returned by API functions for errors and reports them to
+the console. It's not essential to the API.
```
It is not essential to the API but it is a good practice to check the error codes of the HIP APIs incase you pass on incorrect values to the API or the API might be out of resources.

neon60 replied:
Added.

cjatin commented on docs/tutorial/saxpy.rst:
```diff
@@ -0,0 +1,780 @@
+This function is launched from the host using a language extension often called
+the triple chevron syntax. Inside the angle brackets, provide the following.
+
+- The number of blocks to launch (our grid size)
```
Do we have any documentation describing grid and block sizes? If yes can we point the user to it from here. I think these are non trivial concepts and user who just wants to exec stuff on their GPU can go it it if they want to.

neon60 replied:
Added.

cjatin commented on docs/tutorial/saxpy.rst:
```diff
@@ -0,0 +1,780 @@
+- A check is made to avoid overindexing the input.
+- The useful part of the computation is carried out.
+
+Retrieval of the result from the device is done much like its dispatching:
```
dispatching can mean different thing in context of HIP, can we change the wording here. Also the first memcpy should also be explained and here we can mention in simple terms that we are copying result back to host.

neon60 replied:
Removed dispatching.

cjatin commented on docs/tutorial/saxpy.rst:
```diff
@@ -0,0 +1,780 @@
+
+      clang++ .\HIP-Basic\saxpy\main.hip -o saxpy.exe -I .\Common -lamdhip64 -L ${env:HIP_PATH}lib -O2
+
+  .. tab-item:: Windows and NVIDIA
```
Have you tested this particular config? windows + nvidia?

neon60 replied:
As I know, all the commands were tested. I can test this, today. Do you so problems with it?

cjatin replied:
I never really tested this config so was not sure. If it works I am more than OK with it.

cjatin commented on docs/tutorial/saxpy.rst:
```diff
@@ -0,0 +1,780 @@
+  .. tab-item:: Linux and AMD
+    :sync: linux-amd
+
+    The utilities included with ROCm help significantly to inspect binary
```
I think binary inspection deserves its own doc, rather then putting it alongside saxpy explanation. In my head person looking for saxpy documentation will be least bothered with code object extraction and the person wanting to play with code objects will be least bothered with saxpy.

neon60 replied:
We can put this feedback to a separate issue. Currently, we do not have binary inspection tutorial and at least we have some here now. 

If we finally have the binary inspection page, we can rethink this suggestion.  Maybe I can put this into a dropdown:
https://sphinx-design.readthedocs.io/en/latest/dropdowns.html

What do you think?

cjatin replied:
I would rather have its own page, playing with code objects might require a bit more detail and populating here will overcomplicate this tutorial.

neon60 replied:
OK. I am creating a separate issue to remove the code inspection from here to a new code inspection tutorial. 

The Boy Scout Rule: Leave the campground cleaner than you found it. → I am merging this in. One tutorial better than zero, and we will improve this later.

cjatin reviewed (approved):
Looks good

neon60 commented:
> Looks good

Thanks for the review.

neon60 committed:
Add MathiasMagnus SAXPY tutorial

- Move md files to rst
- Impl editorial suggestions from @lpaoletti
- External PR feedbacks from HIP team
